File: Codes in C++/Binary Tree/Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */ﬁ
class Solution {
public:
    int maxDepth(TreeNode* root,int &diameter) {
        if(!root)
            return 0;
        int hl=maxDepth(root->left,diameter);
        int hr=maxDepth(root->right,diameter);
        diameter=max(diameter,hl+hr);
        return 1+ max(hl,hr);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter=0;
        maxDepth(root,diameter);
        return diameter;
    }
};

File: Codes in Python3/Binary Tree/Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode],diameter) -> int:
        if root==None:
            return 0
        hl=self.maxDepth(root.left,diameter)
        hr=self.maxDepth(root.right,diameter)
        diameter[0]=max(diameter[0],hl+hr)
        return 1+max(hl,hr) 
       
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter=[0]
        self.maxDepth(root,diameter)
        return diameter[0]